### PR TITLE
refactoring stuffs, fix explicit types

### PR DESCRIPTION
**Pull request by artofnothingness**
- move all data-only types to "models" namespace
- add few models
- refactored setSpeedLimit 
- explicit types

### CODE REVIEW

**Comment by artofnothingness** [comment]
I'll merge it temporary, if you have any issue i'll make changes in next PR

File: src/optimizer.cpp [hunk 1]
```diff
@@ -41,24 +41,24 @@ void Optimizer::getParams()
...
-  getParam(control_sequence_shift_offset_, "control_sequence_shift_offset", 1);
+  getParam(default_constraints_.vx, "vx_max", 0.5);
+  getParam(default_constraints_.vy, "vy_max", 1.3);
+  getParam(default_constraints_.vw, "wz_max", 1.3);
+  constraints_ = default_constraints_;
+
+  getParam(sampling_std_.vx, "vx_std", 0.1);
+  getParam(sampling_std_.vy, "vy_std", 0.1);
+  getParam(sampling_std_.wz, "wz_std", 0.3);
+  getParentParam(controller_frequency_, "controller_frequency", 0.0);
```
Comment (SteveMacenski) [inline src/optimizer.cpp:60]: Please revert this to the 

```
  nav2_util::declare_parameter_if_not_declared(
    node, "controller_frequency", rclcpp::ParameterValue(0.0));
  node->get_parameter("controller_frequency", controller_frequency_);
```

block, I'm not sure why, but it was causing testing warnings about temporaries that I didn't fully understand but doing the usual parameter declaration worked well

File: include/mppic/models/sampling_std.hpp [hunk 2]
```diff
@@ -0,0 +1,13 @@
+#pragma once
+
+namespace mppi::models
+{
+
+struct SamplingStd
+{
```
Comment (SteveMacenski) [inline include/mppic/models/sampling_std.hpp:7]: Isn't this still a control constraint? Wouldn't making this struct in the `control_constraints.hpp` file still make sense? I don't like having a bunch of these small files with just struct defs if we can help it.

Also, all these new files needs the copyright line for linters

File: include/mppic/optimizer.hpp [hunk 3]
```diff
@@ -17,18 +18,22 @@
...
+#include "mppic/models/control_sequence.hpp"
+#include "mppic/models/control_constraints.hpp"
+#include "mppic/models/state.hpp"
+#include "mppic/models/sampling_std.hpp"
 #include "mppic/utils.hpp"
 
 namespace mppi
 {
 
+using StampType = decltype(std::declval<std_msgs::msg::Header>().stamp);
```
Comment (SteveMacenski) [inline include/mppic/optimizer.hpp:30]: This is a very odd line. Why not use the actual type? https://github.com/ros2/common_interfaces/blob/master/std_msgs/msg/Header.msg#L6

File: include/mppic/models/sampling_std.hpp (hunk 2 above)
Comment (SteveMacenski) [reply include/mppic/models/sampling_std.hpp:7]: Or even put the stds inside of the constraints struct, since that is part of the constraint

File: src/controller.cpp [hunk 4]
```diff
@@ -88,23 +83,21 @@ void Controller::setPlan(const nav_msgs::msg::Path & path)
 
 void Controller::setSpeedLimit(const double & speed_limit, const bool & percentage)
 {
-  utils::ControlConstraints constraints;
+  models::ControlConstraints constraints = optimizer_.getDefaultControlConstraints();
 
-  if (speed_limit == nav2_costmap_2d::NO_SPEED_LIMIT) {
-    // Restore default value
-    constraints = utils::ControlConstraints{base_x_vel_, base_y_vel_, base_theta_vel_};
```
Comment (SteveMacenski) [inline src/controller.cpp:95]: You need this line, or else if you set the speed limit and then you set it to `NO_SPEED_LIMIT` it will not revert back to the base value.

You removed the check for if its `==` to the no speed limit in order to reset the behavior back to default

File: src/controller.cpp [hunk 5]
```diff
@@ -88,23 +83,21 @@ void Controller::setPlan(const nav_msgs::msg::Path & path)
...
-  } else {
+  if (speed_limit != nav2_costmap_2d::NO_SPEED_LIMIT) {
     if (percentage) {
       // Speed limit is expressed in % from maximum speed of robot
-      constraints.vx = base_x_vel_ * speed_limit / 100.0;
-      constraints.vy = base_y_vel_ * speed_limit / 100.0;
-      constraints.vw = base_theta_vel_ * speed_limit / 100.0;
+      double ratio = speed_limit / 100.0;
+      constraints.vx *= ratio;
```
Comment (SteveMacenski) [inline src/controller.cpp:92]: This is wrong as well, this will apply the change on the previously applied change, this must be done using the base values

File: src/controller.cpp [hunk 6]
```diff
@@ -88,23 +83,21 @@ void Controller::setPlan(const nav_msgs::msg::Path & path)
...
+      constraints.vy *= ratio;
+      constraints.vw *= ratio;
     } else {
       // Speed limit is expressed in absolute value
-      double ratio = speed_limit / base_x_vel_;
+      double ratio = speed_limit / constraints.vx;
```
Comment (SteveMacenski) [inline src/controller.cpp:97]: This is wrong too

File: src/controller.cpp [hunk 7]
```diff
@@ -88,23 +83,21 @@ void Controller::setPlan(const nav_msgs::msg::Path & path)
...
       constraints.vx = speed_limit;
-      constraints.vy = base_y_vel_ * ratio;
-      constraints.vw = base_theta_vel_ * ratio;
+      constraints.vy *= ratio;
```
Comment (SteveMacenski) [inline src/controller.cpp:99]: Same reasons this is wrong

**Comment by SteveMacenski** [comment]
@artofnothingness There are critical errors in this PR, I would suggest reverting it or fixing ASAP.

I also really do not like the change from `base` velocities to "default" in the other class, this is not how any other controller in the Nav2 ecosystem works, so I think all of that should be immediately reverted too for consistency. They are not default values, they are base reference values. That verbage doesn't describe what they are and is different from every other controller in Nav2's naming convention and location.

File: include/mppic/optimizer.hpp (hunk 3 above)
Comment (artofnothingness) [reply include/mppic/optimizer.hpp:30]: Oh, sure, i'll make changes in next PR

File: src/controller.cpp (hunk 4 above)
Comment (artofnothingness) [reply src/controller.cpp:95]: it gets default values in first lines of the func, so if it's NO_SPEED_LIMIT he just get and set defaults to the current constraints

File: src/controller.cpp (hunk 5 above)
Comment (artofnothingness) [reply src/controller.cpp:92]: as i mentioned above, constraints already get default values, so we can use them

File: src/controller.cpp (hunk 4 above)
Comment (SteveMacenski) [reply src/controller.cpp:95]: Oh I see, sorry I missed that line in the diff

**Comment by SteveMacenski** [comment]
While I'm not in love with the change in the constraints, I think its OK if the "default" naming goes back to "base" with the refactored API / variable names. That's my biggest hiccup, so that someone that is familiar with any of the controllers can immediately read this and know what they're looking at

File: include/mppic/models/sampling_std.hpp (hunk 2 above)
Comment (artofnothingness) [reply include/mppic/models/sampling_std.hpp:7]: This struct consist of parameters for sampling strategy (std of gaussian distribution).  Semantically very different for me, but i could be wrong. We could have in future different sampling strategies, so why not separate concerns ?

Comment (SteveMacenski) [reply include/mppic/models/sampling_std.hpp:7]: Even if there are future "constraints" of various forms, I think they're still constraints to belong in that struct. I think it would make sense to group them together

Comment (artofnothingness) [reply include/mppic/models/sampling_std.hpp:7]: Constraints impose limits on state vars. This gaussian standard deviation set up sampling strategy params / optimization problem search params. std does not constrain final state vars, it changes how we search result vars, so  its by definition seems to me different, although it limits somehow variables, but it's more like сonsequence.
Control constraints is operationally useful with setSpeedLimit controller API, if we make one general constraint struct, we would pass that there too for no reason. For this API we need 3 double struct to make interface clear. For base_constraints/base_speed_limits, and current constraints/speed_limits  

If I'm not convinced you xD 
We can do something that you propose. But for clear API with setSpeedLimit we need struct that represents that 3 base velocitites and 3 current velocities

Comment (SteveMacenski) [reply include/mppic/models/sampling_std.hpp:7]: I don't think that's unreasonable way of thinking, but complimentary, I think having 2 files with just a 3 `double` struct is silly and I think they have natural grouping as max velocity in XYW and the distributions in XYW :laughing: Perhaps a different word than 'constraint' could be used that would help.

Though, I do think the std's are constraints. Not control constraints, but sampling constraints. 

Maybe even just having a single `constraints.hpp` file with the 2 structs: `control constraints` and `sampling constraints`?

Comment (artofnothingness) [reply include/mppic/models/sampling_std.hpp:7]: > Though, I do think the std's are constraints. Not control constraints, but sampling constraints.

I just try to stick to more academic terms xD I've never heard that std is gaussian distribution constraint (May be i'm just not aware). Constraints for me more about optimization field term. 

> I think having 2 files with just a 3 double struct is silly

I don't like that much too. But choosing between single responsibility principle + modularity and a lot of files, i think first one is better. Anyhow i will place them in single constraints file, if it feels more comfortable, and we could continue discuss design in next PR

File: src/optimizer.cpp (hunk 1 above)
Comment (artofnothingness) [reply src/optimizer.cpp:60]: I found a way to get around it

**Comment by SteveMacenski** [comment]
@artofnothingness changing Default -> Base again would be great

tl;dr
- [x] Single file for both constraint structs
- [x] Using the header timestamp type
- [x] Rename default -> base again
